Cubical/M-types/helper.agda

```diff
@@ -0,0 +1,317 @@
+    subst Y (refl) y
+      ≡⟨ substRefl {B = Y} y ⟩
+    y ∎))
+
+Σ-ap₁ : ∀ {i} {X X' : Set i} {Y : X' → Set i}
+          → (isom : X ≡ X')
+          → Σ X (Y ∘ transport isom) ≡ Σ X' Y
+Σ-ap₁ {i} {X = X} {X'} {Y} isom = isoToPath (Σ-ap-iso₁ (pathToIso isom))
+
+Σ-ap-iso : ∀ {i} {X X' : Set i}
+           {Y : X → Set i} {Y' : X' → Set i}
+         → (isom : Iso X X')
+         → ((x : X) → Iso (Y x) (Y' (fun isom x)))
+         → Iso (Σ X Y)
+                (Σ X' Y')
+Σ-ap-iso {X = X} {X'} {Y} {Y'} isom isom' = compIso (Σ-ap-iso₂ isom') (Σ-ap-iso₁ isom)
+
+Σ-ap :
+  ∀ {i} {X X' : Set i} {Y : X → Set i} {Y' : X' → Set i}
+  → (isom : X ≡ X')
+  → ((x : X) → Y x ≡ Y' (transport isom x))
+  ----------
+  → (Σ X Y)
+  ≡ (Σ X' Y')
+Σ-ap  {X = X} {X'} {Y} {Y'} isom isom' = isoToPath (Σ-ap-iso (pathToIso isom) (pathToIso ∘ isom'))
+
+Σ-ap' :
+  ∀ {ℓ} {X X' : Set ℓ} {Y : X → Set ℓ} {Y' : X' → Set ℓ}
+  → (isom : X ≡ X')
+  → (PathP (λ i → isom i → Set _) Y Y')
+  ----------
+  → (Σ X Y)
+  ≡ (Σ X' Y')
+Σ-ap'  {ℓ} {X = X} {X'} {Y} {Y'} isom isom' = cong₂ (λ (a : Set ℓ) (b : a → Set ℓ) → Σ a λ x → b x) isom isom'
+
+---
+
+identity-x : ∀ {ℓ} {A B : Set ℓ} (k : A -> A) -> k ≡ idfun A -> ∀ (x : A) -> k x ≡ x
+identity-x {A = A} k = funExt⁻
+
+-- Right
+extent-r : ∀ {ℓ} {A B C : Set ℓ} {a b : A -> B} (f : C -> A) -> a ≡ b -> a ∘ f ≡ b ∘ f
+extent-r = λ f x i → x i ∘ f
+
+identity-f-r : ∀ {ℓ} {A B : Set ℓ} {k : A -> A} -> k ≡ idfun A -> ∀ (f : B -> A) -> k ∘ f ≡ f
+identity-f-r {A = A} {k = k} p f = extent-r {a = k} {b = idfun A} f p
+
+-- Left
+extent-l : ∀ {ℓ} {A B C : Set ℓ} {a b : A -> B} (f : B -> C) -> a ≡ b -> f ∘ a ≡ f ∘ b
+extent-l = λ f x i → f ∘ x i
+
+identity-f-l : ∀ {ℓ} {A B : Set ℓ} {k : A -> A} -> k ≡ idfun A -> ∀ (f : A -> B) -> f ∘ k ≡ f
+identity-f-l {A = A} {k = k} p f = extent-l {a = k} {b = idfun A} f p
+
+-- General
+
+≡-rel-a-monomorphism : ∀ {ℓ} {A B C : Set ℓ} (a : A -> B) (b : B -> A) -> a ∘ b ≡ idfun B -> b ∘ a ≡ idfun A -> {f g : C -> A} -> (a ∘ f ≡ a ∘ g) -> (f ≡ g)
+≡-rel-a-monomorphism a b left right {f = f} {g = g} p = λ i →
+  compPath-filler {x = f} {y = (b ∘ a ∘ f)} {z = g}
+    (sym (identity-f-r right f))
+    (λ j → compPath-filler {y = b ∘ a ∘ g}
+      (λ k → b ∘ p k)
+      (identity-f-r right g)
+        j j)
+      i i
+
+transport-iso :
+  ∀ {ℓ} {X Y : Set ℓ}
+  → (isom : Iso X Y)
+  → transport (isoToPath isom) ≡ fun isom
+transport-iso isom = funExt (transportRefl ∘ fun isom)
+
+≡-to-embedding : ∀ {ℓ} {A B C : Set ℓ}
+  → (isom : Iso A B)
+  -------------------------------
+  → isEmbedding (fun isom)
+≡-to-embedding {A = A} {B} {C} isom = (isEquiv→isEmbedding (equivIsEquiv (isoToEquiv isom)))
+
+funExtIso : ∀ {ℓ ℓ₁} {A : Type ℓ} {B : A → Type ℓ₁} {f g : (x : A) → B x} → Iso (∀ x → f x ≡ g x) (f ≡ g)
+funExtIso = iso funExt funExt⁻ refl-fun refl-fun
+
+≡-rel-a-inj' : ∀ {ℓ} {A B C : Set ℓ} (a : A -> B) (e : isEmbedding a) → ∀ {f g : C -> A} -> ∀ x → ((a (f x) ≡ a (g x)) ≡ (f x ≡ g x))
```

Comment:
What does `≡-rel` mean in this context? It's used in a lot of the lemmas in the rest of this file.

Reply:
I don't think the names are very good currently, I wrote `≡-rel` to mean some equivalence relation.